Enable Integration Tests for Jörmungandr (Part III: Command-Line tests)

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#358 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have introduce a new class-constraints `KnownCommand` to allow abstracting the name of the command-line interface from the CLI DSL and scenarios
- [x] I have renamed `cardano-wallet` executable into `cardano-wallet-http-bridge`
- [x] I isolated the `Server` specs for `http-bridge` and created a specific one for Jormungandr
- [x] I have enabled most of the integration tests specs, except the Ports one (require a slightly more complex setup)

To be done:

- [ ] Enable the Port tests
- [ ] Review `.travis.yaml` deployment script


# Comments

<!-- Additional comments or screenshots to attach if any -->

:warning: Base branch is `KtorZ/cli-optparse-applicative` :warning: 

**spoiler:** Most of this PR is about `@t`.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
remove use of 'generateMnemonicsViaCLI' from API scenarios

KtorZ committed:
define command-line name through a type constraint via 'KnownCommand'

This allows for running the same scenarios using different command line interfaces

KtorZ committed:
name 'cardano-wallet' executable to 'cardano-wallet-http-bridge' and implement 'KnownCommand' for it

KtorZ committed:
enable (almost all) CLI integration tests for Jörmungandr

Remains: the port tests which are a bit trickier. Will do in a separate commit

KtorZ committed:
fix oversight about amount comparison for multiple output CLI transaction

KtorZ commented on lib/core/test/integration/Test/Integration/Scenario/API/Wallets.hs:
```diff
@@ -134,8 +133,8 @@ spec = do
-        Stdout mnemonics <- generateMnemonicsViaCLI []
-        let payldCrt = payloadWith "!st created" (T.words $ T.pack mnemonics)
+        mnemonics <- mnemonicToText @15 . entropyToMnemonic <$> genEntropy
+        let payldCrt = payloadWith "!st created" mnemonics
```
I've "inlined" this CLI call since there's no reason the API scenarios would require something from the CLI (and this complexifies calling these specs for now reasons). We do have dedicated CLI specs for testing mnemonic generation so, all good.

KtorZ commented on lib/http-bridge/test/integration/Main.hs:
```diff
@@ -89,59 +92,63 @@ import qualified Data.Aeson.Types as Aeson
+instance KnownCommand (HttpBridge n) where
+    commandName = "cardano-wallet-http-bridge"
+
+main :: forall t. (t ~ HttpBridge 'Testnet) => IO ()
```
And this is where the magic happens and where we finally get to choose which CLI the specs should run on.

KtorZ commented on lib/core/test/integration/Test/Integration/Scenario/CLI/Transactions.hs:
```diff
@@ -204,18 +207,18 @@ spec = do
         txJson <- expectValidJSON (Proxy @(ApiTransaction t)) out
         verify txJson
-            [ expectCliFieldBetween amount (feeMin + amt, feeMax + amt)
+            [ expectCliFieldBetween amount (feeMin + (2*amt), feeMax + (2*amt))
```
So, this was left unnoticed because the fee amounts are much bigger on the bridge scenarios and make the amount completely insignificant. This isn't the case with Jörmungandr where the amount and fee are more of less of the same order of magnitude!

jonathanknowles committed:
Make nix happy again.

rvl commented on lib/core/test/integration/Test/Integration/Framework/DSL.hs:
```diff
@@ -838,28 +850,47 @@ createWalletViaCLI ctx args mnemonics secondFactor passphrase = do
-deleteWalletViaCLI :: (CmdResult r, HasType Port s) => s -> String -> IO r
-deleteWalletViaCLI ctx walId = cardanoWalletCLI
+deleteWalletViaCLI
+    :: forall t r s. (CmdResult r, KnownCommand t, HasType Port s)
```
Maybe we could use a constraint synonym to reduce noise.

KtorZ replied:
Yeah, thought of it but some functions have slightly different constraints, and I thought that introducing sometimes an alias, and sometimes not will create more noise than help. In the end, it will mean one other type of constraint to load into our brain when reading :/

rvl commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Server.hs:
```diff
@@ -0,0 +1,44 @@
+                let process = proc' (commandName @t) args
+                withCreateProcess process $ \_ _ _ ph -> do
+                    expectPathEventuallyExist db
+                    expectPathEventuallyExist (db <> "-shm")
```
I know this is copy&pasted from somewhere else, but will comment here anyway.

I don't think we need to assert that the `-shm` and `-wal` files exist. They are meant to be [temporary files](https://www.sqlite.org/tempfiles.html), and are a sqlite implementation detail. We just need to assert the db file was created.

Also if the `-wal` file exists after exit then that means we have not closed the database connection properly. Hmm.

KtorZ replied:
I asked a question in the past about this on a review (about what where those -shm and -wal expected to be). I agree that we could reduce this to test whether the db file exists as the rest seems more like an "implementation detail" from SQLite & Persistent (as far as I understand...)

KtorZ committed:
remove assertion on temporary db files

Rationale is that these files are rather internal to the way the database works. So, there's no
reason we should be even considering them. As long as the db file exists, we are fine.
